proposals/defer-hydration.md

```diff
@@ -0,0 +1,90 @@
+# Defer Hydration Protocol
+
+An open protocol for controlling hydration on the client.
+
+Author: Justin Fagnani
+
+Status: Draft
+
+Last update: 2021-06-24
+
+# Background
+
+In server-side rendered (SSR) applications, the process of a component running code to re-associate its template with the server-rendered DOM is called "hydration". The defer-hydration protocol is design to allow controler over hydration to solve two related problems:
+
+1. Interoperable incremental hydration across web components. Componentents should not automatically hydrate upon being defined, but wait for a signal from their parent or other coordinator.
+2. Hydration ordering independent of definition order. Because components usually depend on data from their parent, and the parent won't usually set data on a child until it's hydrated, we need hydration to occur in a top-down order.
+
+`defer-hydration` enables us to decouple loading the code for web component definitions from starting the work of hydration, and enables top-down ordering and sophisticated coordination of hydration, including triggering hydration only on interaction or data changes for specific components.
+
+# Overview
+
+The Defer Hydration Protocol specifies an attribute named `defer-hydration` that is placed on elements, usually during server rendering, to tell them not to hydrate when they are upgraded. Removing this attribute is a signal that the element should hydrate. Elements can observe the attribute being removed via `observedAttribute` and `attributeChangedCallback`.
```

Comment:
Curious on folks perspective around `defer-hydration` as compared to something akin to `loading="lazy"`

If the attribute was, say, `hydration`, with options like: 
 - `defer`: Hydrates based on internal behavior
 - `lazy`: Hydrates based on when in viewport
 - `eager`: Default behavior
 
Should the attribute control both the intended behavior and the current state? Some tools use `:visible`, Stencil uses the `hydrated` class for hydrated state. 

Decoupling this and setting a convention around the intended behavior and leaving how state is communicated up to the tools might be helpful.